### PR TITLE
User/jherrild/test

### PR DESCRIPTION
**Pull request by Jherrild**
1. **What this PR does / why we need it**:

Adds acceptance test for eks cluster creation/deletion

3. **Which issue(s) this PR fixes**

4. **Additional information**

Uses additional environment variable 'ENABLE_EKS_ENV_TEST' to enable actual calls against MKP stack- otherwise runs with locally mocked HTTP calls.

5. **Special notes for your reviewer**

### CODE REVIEW

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 1]
```diff
@@ -550,8 +559,23 @@ func handleClusterDiff(config authctx.TanzuContext, tmcCluster *eksmodel.VmwareT
 		tmcCluster.Meta.Labels = meta.Labels
 	}
 
-	if !reflect.DeepEqual(clusterSpec, tmcCluster.Spec) {
-		updateCluster = true
+	// Check for equality of spec- ignore differences relating to tags generated by TMC
+	if diff := deep.Equal(clusterSpec, tmcCluster.Spec); len(diff) != 0 {
```
Comment (asutoshpalai) [inline internal/resources/ekscluster/resource_ekscluster.go:563]: Why would these tags change in the first place?

File: internal/resources/testing/constant.go [hunk 2]
```diff
@@ -14,3 +14,5 @@ const (
...
...:iam::919197287370:role/worker.TEST_CLOUD_FORMATION_TEMPLATE_ID.eks.tmc.cloud.vmware.com\",\"rootDiskSize\":40,\"scalingConfig\":{\"desiredSize\":4,\"maxSize\":8,\"minSize\":1},\"subnetIds\":[\"subnet-0ed95d5c212ac62a1\",\"subnet-06897e1063cc0cf4e\",\"subnet-0526ecaecde5b1bf7\",\"subnet-0a184f6302af32a86\"],\"tags\":{\"testnptag\":\"testnptagvalue\"},\"taints\":[],\"updateConfig\":{\"maxUnavailableNodes\":\"2\"}}},{\"info\":{\"description\":\"tf nodepool 2 description\",\"name\":\"second-np\"},\"spec\":{\"instanceTypes\":[],\"launchTemplate\":{\"name\":\"vivek\",\"version\":\"7\"},\"nodeLabels\":{\"testnplabelkey\":\"testnplabelvalue\"},\"remoteAccess\":{\"securityGroups\":null},\"roleArn\":\"arn:aws:iam::919197287370:role/worker.TEST_CLOUD_FORMATION_TEMPLATE_ID.eks.tmc.cloud.vmware.com\",\"scalingConfig\":{\"desiredSize\":4,\"maxSize\":8,\"minSize\":1},\"subnetIds\":[\"subnet-0ed95d5c212ac62a1\",\"subnet-06897e1063cc0cf4e\",\"subnet-0526ecaecde5b1bf7\",\"subnet-0a184f6302af32a86\"],\"tags\":{\"testnptag\":\"testnptagvalue\"},\"taints\":[{\"effect\":\"PREFER_NO_SCHEDULE\",\"key\":\"randomkey\",\"value\":\"randomvalue\"}],\"updateConfig\":{\"maxUnavailablePercentage\":\"12\"}}}]}}}"
```
Comment (asutoshpalai) [inline internal/resources/testing/constant.go:18]: nit: is it possible to paste the formatted JSON here for easy readability? Or create the original struct and convert it to JSON on the fly/statically?

File: internal/resources/ekscluster/resource_ekscluster.go (hunk 1 above)
Comment (Jherrild) [reply internal/resources/ekscluster/resource_ekscluster.go:563]: This was logic I added to compensate for us adding the TMC tags that aren't in the original terraform spec.

File: internal/resources/testing/constant.go (hunk 2 above)
Comment (Jherrild) [reply internal/resources/testing/constant.go]: This actually isn't used anymore- thanks for flagging it! Removed!

File: go.mod [hunk 3]
```diff
@@ -15,6 +16,8 @@ require (
 	sigs.k8s.io/controller-runtime v0.6.0
 )
 
+require github.com/google/go-cmp v0.5.8 // indirect
```
Comment (shreyassreenivas) [inline go.mod:19]: why is this require in a separate block ?

File: internal/authctx/helper.go [hunk 4]
```diff
@@ -94,6 +95,32 @@ func ProviderConfigureContext(_ context.Context, d *schema.ResourceData) (interf
 	return setContext(&config)
 }
 
+func ProviderConfigureContextWithDefaultTransport(_ context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
```
Comment (shreyassreenivas) [inline internal/authctx/helper.go:98]: can you explain why is this DefaultTransport needed ?

File: internal/client/http_client.go [hunk 5]
```diff
@@ -38,6 +38,15 @@ func NewHTTPClient(config *proxy.TLSConfig) (*TanzuMissionControl, error) {
 		return nil, err
 	}
 
+	return newHTTPClient(httpClient), nil
+}
+
+// NewHTTPClientWithDefaultTransport is intended primarily for testing, as httpmock requires a default transport object be used.
+func NewHTTPClientWithDefaultTransport() *TanzuMissionControl {
```
Comment (shreyassreenivas) [inline internal/client/http_client.go:45]: if it is going to be used only for tests, can this be added only in the test files and not in the main code flow ?

File: internal/resources/ekscluster/resource_ekscluster_test.go [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+				// Then, sort slices and compare
+				for _, set := range subnetMap {
+					left := set["left"]
+					right := set["right"]
+
+					// sort
+					sort.Strings(left)
+					sort.Strings(right)
+
+					subnetDiff := deep.Equal(left, right)
```
Comment (shreyassreenivas) [inline internal/resources/ekscluster/resource_ekscluster_test.go:578]: can you use the deep.Equal from the testing/require package ? 
what is the advantage of this new deep package ?

File: internal/resources/testing/test_helper.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		ResourceNameVar:          EksClusterResourceVar,
+		Meta:                     MetaTemplate,
+		AccTestType:              CreateEksCluster,
+		TemplateData:             testDefaultCreateEksClusterScript,
+		OrgID:                    "bc27608b-4809-4cac-9e04-778803963da2",
+		AWSAccountNumber:         "919197287370",
+		Region:                   "us-west-2",
+		ClusterGroupName:         "default",
+		KubernetesVersion:        "1.23",
+		LaunchTemplateName:       "vivek",
```
Comment (shreyassreenivas) [inline internal/resources/testing/test_helper.go:197]: can we avoid using names ?

File: internal/resources/testing/test_config.go [hunk 8]
```diff
@@ -3,7 +3,129 @@ Copyright © 2022 VMware, Inc. All Rights Reserved.
...
+					tags           = { "testnptag" : "testnptagvalue" }
+					node_labels    = { "testnplabelkey" : "testnplabelvalue" }
+					subnet_ids = [ // Required, forces new
+						"subnet-0a184f6302af32a86",
+						"subnet-0ed95d5c212ac62a1",
+						"subnet-0526ecaecde5b1bf7",
+						"subnet-06897e1063cc0cf4e",
+					]
+					remote_access  {     // Forces new
+						ssh_key = "anshulc" // Required (for remote access)
```
Comment (shreyassreenivas) [inline internal/resources/testing/test_config.go:69]: avoid using names

File: internal/resources/ekscluster/resource_ekscluster.go [hunk 9]
```diff
@@ -132,6 +140,7 @@ var configSchema = &schema.Schema{
 				Description: "The metadata to apply to the cluster to assist with categorization and organization",
 				Optional:    true,
 				Elem:        &schema.Schema{Type: schema.TypeString},
+				Computed:    true,
```
Comment (shreyassreenivas) [inline internal/resources/ekscluster/resource_ekscluster.go:143]: is this intended ?, User will not be able to provide values if marked as computed

File: internal/resources/ekscluster/resource_ekscluster.go (hunk 1 above)
Comment (shreyassreenivas) [reply internal/resources/ekscluster/resource_ekscluster.go:563]: can we use the suppress diff function which is being used in other places ?

File: go.mod [hunk 10]
```diff
@@ -5,6 +5,7 @@ go 1.18
 require (
 	github.com/go-openapi/strfmt v0.19.3
 	github.com/go-openapi/swag v0.19.5
+	github.com/go-test/deep v1.0.3
```
Comment (shreyassreenivas) [inline go.mod:8]: I see that the go-test/deep is being used in non test files, can this be avoided  ?

Comment (Jherrild) [reply go.mod:8]: Sure, thanks for pointing this out, I will change it

File: internal/authctx/helper.go (hunk 4 above)
Comment (Jherrild) [reply internal/authctx/helper.go:98]: The default transport is needed for mocking- the http mocking library I'm using can only intercept calls if they're made with the default transport.

File: go.mod (hunk 3 above)
Comment (Jherrild) [reply go.mod:19]: I have no idea- I didn't add it manually. I'll try merging them.

File: internal/resources/ekscluster/resource_ekscluster.go (hunk 1 above)
Comment (Jherrild) [reply internal/resources/ekscluster/resource_ekscluster.go]: Removed

File: go.mod (hunk 3 above)
Comment (Jherrild) [reply go.mod]: Fixed

File: internal/resources/ekscluster/resource_ekscluster_test.go (hunk 6 above)
Comment (Jherrild) [reply internal/resources/ekscluster/resource_ekscluster_test.go:582]: It returns a human readable diff to be compared- I will attempt to replace this tomorrow.

File: internal/resources/testing/test_helper.go (hunk 7 above)
Comment (Jherrild) [reply internal/resources/testing/test_helper.go]: Removed

File: internal/client/http_client.go (hunk 5 above)
Comment (shreyassreenivas) [inline internal/client/http_client.go:45]: Can we have the test word added in the function name itself

**Comment by Jherrild** [comment]
Remote test (against real MKP)
image

Local test (mocked HTTP)
image

File: internal/client/http_client.go (hunk 5 above)
Comment (Jherrild) [reply internal/client/http_client.go]: Added!

**Pull request merged**